documents/nft_created_task.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/centrifuge/gocelery"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/satori/go.uuid"
+)
+
+const (
+	// nftCreatedTaskName is the name of the NFT Task
+	nftCreatedTaskName string = "NFT created task"
+
+	// TokenRegistryParam maps to token registry address
+	TokenRegistryParam string = "Token Registry"
+
+	// TokenIDParam maps to NFT token ID
+	TokenIDParam string = "Token ID"
+)
+
+type nftCreatedTask struct {
+	transactions.BaseTask
+	accountID     identity.CentID
+	documentID    []byte
+	tokenRegistry common.Address
+	tokenID       []byte
+
+	// state
+	docSrv Service
+	cfgSrv config.Service
+}
+
+func (t *nftCreatedTask) ParseKwargs(kwargs map[string]interface{}) (err error) {
+	err = t.ParseTransactionID(kwargs)
+	if err != nil {
+		return err
+	}
+
+	acc, ok := kwargs[AccountIDParam].(string)
+	if !ok {
+		return errors.New("account identifier not set")
+	}
+
+	t.accountID, err = identity.CentIDFromString(acc)
+	if err != nil {
+		return errors.New("invalid accountID")
+	}
+
+	vr, ok := kwargs[DocumentIDParam].(string)
+	if !ok {
+		return errors.New("model identifier not set")
+	}
+
+	t.documentID, err = hexutil.Decode(vr)
+	if err != nil {
+		return err
+	}
+
+	tr, ok := kwargs[TokenRegistryParam].(string)
+	if !ok {
+		return errors.New("token registry not set")
+	}
+
+	t.tokenRegistry = common.HexToAddress(tr)
+
+	tidr, ok := kwargs[TokenIDParam].(string)
+	if !ok {
+		return errors.New("token ID not set")
+	}
+
+	t.tokenID, err = hexutil.Decode(tidr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *nftCreatedTask) Copy() (gocelery.CeleryTask, error) {
+	return &nftCreatedTask{
+		BaseTask: t.BaseTask,
+		docSrv:   t.docSrv,
+		cfgSrv:   t.cfgSrv,
+	}, nil
+}
+
+func (t *nftCreatedTask) RunTask() (result interface{}, err error) {
+	defer func() {
+		err = t.UpdateTransaction(t.accountID, t.TaskTypeName(), err)
+	}()
+
+	ctx := context.WithValue(context.Background(), config.AccountHeaderKey, t.accountID.String())
+	ctx, err = contextutil.Context(ctx, t.cfgSrv)
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := t.docSrv.GetCurrentVersion(ctx, t.documentID)
+	if err != nil {
+		return nil, err
+	}
+
+	cd, err := model.PackCoreDocument()
+	if err != nil {
+		return nil, err
+	}
+
+	data := cd.EmbeddedData
+	cd, err = coredocument.PrepareNewVersion(*cd, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	cd.EmbeddedData = data
+	err = coredocument.AddNFTToReadRules(cd, t.tokenRegistry, t.tokenID)
+	if err != nil {
+		return nil, err
+	}
+
+	model, err = t.docSrv.DeriveFromCoreDocument(cd)
+	if err != nil {
+		return nil, err
+	}
+
+	_, _, err = t.docSrv.Update(ctx, model, t.TxID)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (t *nftCreatedTask) TaskTypeName() string {
+	return nftCreatedTaskName
+}
+
+// InitNFTCreatedTask initiates a new nft created task
+func InitNFTCreatedTask(
+	queuer queue.TaskQueuer,
```

Comment:
Why are you not creating an own txID for the next task in the queue?

Multiply queuing task are using the same txID?

Reply:
This was something we have discussed about having same transaction I'd for multiple tasks. Else we will not be able to communicate the next yxid to the client